In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# BASIC DATA EXPLORATION

import pandas as pd

# Step 1. Loading Data

# Path of the file to read

# NO CAMBIAR EL PATH
path = "/content/drive/MyDrive/Colab Notebooks/EALC/"

iowa_file_path = path + "train.csv"

# Read the file into a variable home_data
home_data = pd.read_csv(iowa_file_path)


# Step 2. Review the data

# Print summary statistics in next line
home_data.describe()


Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

### Material del curso Intro to Machine Learning de Kaggle

### https://www.kaggle.com/learn/intro-to-machine-learning

# Basic data exploration


In [3]:
# What is the average lot size (rounded to nearest integer)?
avg_lot_size = home_data['LotArea'].mean().round()
print("Average lot size:", int(avg_lot_size))

# As of today, how old is the newest home (current year - the date in which it was built)
current_year = 2023
newest_home_age = (current_year - home_data['YearBuilt']).min()
print("La casa más nueva tiene", newest_home_age, "años.")

Average lot size: 10517
La casa más nueva tiene 13 años.


# Your first machine learning model

### Step 1: Specify prediction target

In [4]:
# print the list of columns in the dataset to find the name of the prediction target
print(home_data.columns)

y = home_data.SalePrice

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

### Step 2: create X

In [6]:
# Create the list of features below
feature_names = ["LotArea", "YearBuilt", "1stFlrSF", "2ndFlrSF", "FullBath", "BedroomAbvGr", "TotRmsAbvGrd"]

# Select data corresponding to features in feature_names
X = home_data[feature_names]

Review data

In [7]:
# Review data

# print description or statistics from X
print(X.describe())

# print the top few lines
print(X.head())

             LotArea    YearBuilt     1stFlrSF     2ndFlrSF     FullBath  \
count    1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean    10516.828082  1971.267808  1162.626712   346.992466     1.565068   
std      9981.264932    30.202904   386.587738   436.528436     0.550916   
min      1300.000000  1872.000000   334.000000     0.000000     0.000000   
25%      7553.500000  1954.000000   882.000000     0.000000     1.000000   
50%      9478.500000  1973.000000  1087.000000     0.000000     2.000000   
75%     11601.500000  2000.000000  1391.250000   728.000000     2.000000   
max    215245.000000  2010.000000  4692.000000  2065.000000     3.000000   

       BedroomAbvGr  TotRmsAbvGrd  
count   1460.000000   1460.000000  
mean       2.866438      6.517808  
std        0.815778      1.625393  
min        0.000000      2.000000  
25%        2.000000      5.000000  
50%        3.000000      6.000000  
75%        3.000000      7.000000  
max        8.000000     14.

### Step 3: Specify and fit model

In [8]:
# from _ import _
from sklearn.tree import DecisionTreeRegressor

#specify the model.
#For model reproducibility, set a numeric value for random_state when specifying the model
iowa_model = DecisionTreeRegressor(random_state=1)

# Fit the model
iowa_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

### Step 4: Make predictions

In [9]:
predictions = iowa_model.predict(X)
print(predictions)

[208500. 181500. 223500. ... 266500. 142125. 147500.]


# Model validation

### Step 1: Split your data

In [10]:
# Import the train_test_split function
from sklearn.model_selection import train_test_split

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 1)

### Step 2: Specify and fit the model

In [11]:
# Specify the model
iowa_model = DecisionTreeRegressor(random_state = 1)

# Fit iowa_model with the training data.
iowa_model.fit(train_X,train_y)

DecisionTreeRegressor(random_state=1)

### Step 3: Make predictions with validation data

In [12]:
# Predict with all validation observations
val_predictions = iowa_model.predict(val_X)

In [13]:
# print the top few validation predictions
print(val_predictions[:5])
# print the top few actual prices from validation data
print(train_y.head().tolist())

[186500. 184000. 130000.  92000. 164500.]
[307000, 223500, 145000, 155000, 140000]


### Step 4: Calcute the mean absolute error in validation data

In [14]:
from sklearn.metrics import mean_absolute_error
val_mae = mean_absolute_error(val_y, val_predictions)

# uncomment following line to see the validation_mae
print(val_mae)


29652.931506849316


# Underfitting and Overfitting

### Exercises

In [15]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

### Step 1: Compare different tree sizes

In [16]:
candidate_max_leaf_nodes = [5, 25, 50, 100, 250, 500]
# Write loop to find the ideal tree size from candidate_max_leaf_nodes
scores = {leaf_size: get_mae(leaf_size, train_X, val_X, train_y, val_y) for leaf_size in candidate_max_leaf_nodes}

# Store the best value of max_leaf_nodes (it will be either 5, 25, 50, 100, 250 or 500)
best_tree_size = min(scores, key=scores.get)
print("Best tree size:", best_tree_size)

Best tree size: 100


### Step 2: Fit model using all data

In [17]:
# Fill in argument to make optimal size and uncomment
final_model = DecisionTreeRegressor(max_leaf_nodes= best_tree_size, random_state=1)

# fit the final model and uncomment the next two lines
final_model.fit(X,y)
predictions_final = final_model.predict(val_X)

val_mae_final = mean_absolute_error(val_y, predictions_final)
print(val_mae_final)

16815.938748057826


# Random forests

In [18]:
from sklearn.ensemble import RandomForestRegressor

# Define the model. Set random_state to 1
rf_model = RandomForestRegressor(random_state=1)

# fit your model
rf_model.fit(train_X, train_y)

# Calculate the mean absolute error of your Random Forest model on the validation data
iowa_preds = rf_model.predict(val_X)
rf_val_mae = mean_absolute_error(val_y, iowa_preds)

print("Validation MAE for Random Forest Model: {}".format(rf_val_mae))

Validation MAE for Random Forest Model: 21857.15912981083


# Preguntas sobre los ejercicios

1. ¿Cuántos datos hay en el dataset inicial?, ¿cuántos datos hay en los conjuntos de entrenamiento y validación que se generan al hacer el split?

In [ ]:
# leer el archivo inicial
dataset_inicio = pd.read_csv(iowa_file_path)

# obtener el número de columnas y filas del dataframe

num_filas, num_columnas = dataset_inicio.shape

print(f'El dataset tiene {num_filas} filas divididas por {num_columnas} columnas.')

El dataset tiene 1460 filas divididas por 81 columnas.


In [ ]:
# renombramos la variable de entrenamiento
entrenamiento = train_X

# obtener el número de filas y columnas del dataframe de entrenamiento
num_filas, num_columnas = entrenamiento.shape

print(f'El dataset de entrenamiento {num_filas} filas divididas por {num_columnas} columnas.')

El dataset de entrenamiento 1095 filas divididas por 7 columnas.


In [ ]:
# renombramos la variable de validación
validacion = val_X

# obtener el número de filas y columnas del dataframe de validación
num_filas, num_columnas = validacion.shape

print(f'El dataset de validacion {num_filas} filas divididas por {num_columnas} columnas.')

El dataset de validacion 365 filas divididas por 7 columnas.


2. Respecto a los modelos desarrollados: ¿cuántos atributos tienen y cuáles son?, ¿qué predicen los modelos?, ¿qué modelo funciona mejor?, ¿qué efectividad tiene?

1. Todos los modelos que hemos desarrollado tienen siete atributos:
- LotArea
- YearBuilt
- 1stFlrSF
- 2ndFlrSF
- FullBath
- BedroomAbvGr
- TotRmsAbvGrd
Lo que cambian son los parámetros de cada modelo

2. Estamos usando modelos de regresión, por tanto, lo que predicen es una variable numérica. En este caso, el precio de las casas en Iowa, que se corresponde con la etiqueta "SalesPrice".

3. Si calculamos el error de los dos modelos que hemos hecho hasta ahora, el que mejor funciona es el DecissionTree con la cantidad de nodos que habíamos considerado más adecuada, en este caso, 100:

In [19]:
print(f'Error del modelo DecisionTree: {val_mae_final}')
print(f'Error del modelo RandomForest: {rf_val_mae}')

Error del modelo DecisionTree: 16815.938748057826
Error del modelo RandomForest: 21857.15912981083


Sin embargo, estos datos no son comparables ya que en el DecisionTree hemos utilizado todos los datos (X,y) para entrenarlo. En el modelo RandomForest solo habíamos usado los datos de entrenamiento (train_X, train_y). Para poder comparar los datos de manera realista, vamos a entrenar al modelo RandomForest con todos los datos:

In [20]:
rf_model_final = RandomForestRegressor(random_state=1)
rf_model_final.fit(X,y)

val_predictions_rf_final = rf_model_final.predict(val_X)
rf_val_mae_final = mean_absolute_error(val_y,val_predictions_rf_final)
print(f'Error del modelo RandomForest total: {rf_val_mae_final}')

Error del modelo RandomForest total: 7879.472059621657


3 y 4. Cuando entrenamos ambos modelos con los mismos datos, vemos que la efectividad del modelo RandomForest es mucho más alta, teniendo un error mucho más bajo que el decisionTree: 7879 frente a 16815.
Por tanto, el modelo que funciona mejor y el más efectivo en el RandomForest.